sdks/java/extensions/sql/src/main/java/org/apache/beam/sdk/extensions/sql/impl/transform/BeamAggregationTransforms.java

```diff
@@ -234,26 +234,29 @@ public AggregationAccumulator mergeAccumulators(Iterable<AggregationAccumulator>
         while (ite.hasNext()) {
           accs.add(ite.next().accumulatorElements.get(idx));
         }
-        deltaAcc.accumulatorElements.add(aggregators.get(idx).merge(accs));
+        deltaAcc.accumulatorElements.add(aggregators.get(idx).mergeAccumulators(accs));
       }
       return deltaAcc;
     }
     @Override
     public BeamRecord extractOutput(AggregationAccumulator accumulator) {
       List<Object> fieldValues = new ArrayList<>(aggregators.size());
       for (int idx = 0; idx < aggregators.size(); ++idx) {
-        fieldValues.add(aggregators.get(idx).result(accumulator.accumulatorElements.get(idx)));
+        fieldValues
+            .add(aggregators.get(idx).extractOutput(accumulator.accumulatorElements.get(idx)));
       }
       return new BeamRecord(finalRowType, fieldValues);
     }
     @Override
     public Coder<AggregationAccumulator> getAccumulatorCoder(
         CoderRegistry registry, Coder<BeamRecord> inputCoder)
         throws CannotProvideCoderException {
+      BeamRecordCoder inCoderAsSql = (BeamRecordCoder) inputCoder;
       registry.registerCoderForClass(BigDecimal.class, BigDecimalCoder.of());
       List<Coder> aggAccuCoderList = new ArrayList<>();
-      for (BeamSqlUdaf udaf : aggregators) {
-        aggAccuCoderList.add(udaf.getAccumulatorCoder(registry));
+      for (int idx = 0; idx < aggregators.size(); ++idx) {
+        aggAccuCoderList.add(aggregators.get(idx).getAccumulatorCoder(registry,
+            inCoderAsSql.getCoderArray().get(sourceFieldExps.get(idx).getInputRef())));
```

Comment:
Can you maybe pull this whole line out into a separate statement and store it in a clearly named variable for use in the getAccumulatorCoder call? This for loop used to read very cleanly, and it's pretty dense now.